src/jit/flowgraph.cpp

```diff
@@ -5746,14 +5746,36 @@ void          Compiler::fgFindBasicBlocks()
 
     if (compIsForInlining())
     {
+        bool hasReturnBlocks = false;
+        bool hasMoreThanOneReturnBlock = false;
+
+        for (BasicBlock* block = fgFirstBB; block; block = block->bbNext)
+        {
+            if (block->bbJumpKind == BBJ_RETURN)
+            {
+                if (hasReturnBlocks)
+                {
+                    hasMoreThanOneReturnBlock = true;
+                    break;
+                }
+
+                hasReturnBlocks = true;
+            }
+        }
+
+        if (!hasReturnBlocks)
+        {
+            impInlineInfo->inlineResult->NoteFatal(InlineObservation::CALLEE_DOES_NOT_RETURN);
+            return;
+        }
+
```

Comment:
Yes, the current inliner -- via LegacyPolicy -- is conservative with respect to size of inlinees. But I'd rather not bake that assumption in here. Going forward, policies will likely be somewhat less conservative. 

It would be simple enough here to just impose a limit on the number of BB's. Or turn this into an informational observation and let the policy decide whether or not to inline, and set the limits there.